TestingLowerBounds/KullbackLeibler.lean

```diff
@@ -453,5 +453,122 @@ lemma kl_fst_add_condKL [StandardBorelSpace β] [Nonempty β] {μ ν : Measure (
+  let s' : (i : ι) → Set (β (e i)) := fun i ↦ s (e i)
+  have : e_meas ⁻¹' Set.pi Set.univ s = Set.pi Set.univ s' := by
+    ext x
+    simp only [Set.mem_preimage, Set.mem_pi, Set.mem_univ, forall_true_left, s']
+    apply (e.forall_congr _).symm
+    intro i
+    convert Iff.rfl
+    have piCongrLeft_apply_apply' :
+        (MeasurableEquiv.piCongrLeft (fun i' => β i') e) x (e i) = x i := by
+      simp only [MeasurableEquiv.piCongrLeft, MeasurableEquiv.coe_mk,
+        Equiv.piCongrLeft_apply_apply]
+    rw [piCongrLeft_apply_apply']
+  rw [this, Measure.pi_pi, Finset.prod_equiv e.symm]
+  · simp only [Finset.mem_univ, implies_true]
+  intro i _
+  simp only [s']
+  congr
+  all_goals rw [e.apply_symm_apply]
+
+lemma kl_pi {ι : Type*} [hι : Fintype ι] {β : ι → Type*} [∀ i, MeasurableSpace (β i)]
+    [∀ i, CountablyGenerated (β i)] {μ ν : (i : ι) → Measure (β i)}
+    [∀ i, IsProbabilityMeasure (μ i)] [∀ i, IsProbabilityMeasure (ν i)] :
+    kl (Measure.pi μ) (Measure.pi ν) = ∑ i, kl (μ i) (ν i) := by
+  refine Fintype.induction_empty_option (P := fun ι ↦ ∀ {β : ι → Type u_4}
+    [(i : ι) → MeasurableSpace (β i)] [∀ (i : ι), CountablyGenerated (β i)]
+    {μ ν : (i : ι) → Measure (β i)} [∀ (i : ι), IsProbabilityMeasure (μ i)]
+    [∀ (i : ι), IsProbabilityMeasure (ν i)],
+    kl (Measure.pi μ) (Measure.pi ν) = ∑ i : ι, kl (μ i) (ν i) ) ?_ ?_ ?_ ι
+  · intro ι ι' hι' e h β _ _ μ ν _ _
+    specialize h (β := fun i ↦ β (e i)) (μ := fun i ↦ μ (e i)) (ν := fun i ↦ ν (e i))
+    let hι : Fintype ι := Fintype.ofEquiv _ e.symm
+    rw [Fintype.sum_equiv e.symm _ (fun i ↦ kl (μ (e i)) (ν (e i))), ← h, kl_eq_fDiv, kl_eq_fDiv]
+    let e_meas : ((b : ι) → β (e b)) ≃ᵐ ((a : ι') → β a) :=
+      MeasurableEquiv.piCongrLeft (fun i ↦ β i) e
+    have me := MeasurableEquiv.measurableEmbedding e_meas.symm
+    convert (fDiv_map_measurableEmbedding me).symm
+    <;> try {rw [← Measure.pi_map_CongrLeft e, MeasurableEquiv.map_symm_map]}
+    <;> infer_instance
+    intro i
+    rw [Equiv.apply_symm_apply]
+  · intro β _ _ μ ν _ _
+    rw [Measure.pi_of_empty, Measure.pi_of_empty, kl_self, Finset.univ_eq_empty, Finset.sum_empty]
+  · intro ι hι ind_h β _ _ μ ν _ _
+    specialize ind_h (β := fun i ↦ β i) (μ := fun i ↦ μ i) (ν := fun i ↦ ν i)
+    have h : kl (Measure.pi μ) (Measure.pi ν) = kl (Measure.prod (Measure.pi (fun (i : ι) ↦ μ i))
+        (μ none)) (Measure.prod (Measure.pi (fun (i : ι) ↦ ν i)) (ν none)) := by
+      rw [kl_eq_fDiv, kl_eq_fDiv]
+      let e_meas : ((i : ι) → β (some i)) × β none ≃ᵐ ((i : Option ι) → β i) :=
+        MeasurableEquiv.piOptionEquivProd.{_, _, u_3} β |>.symm --TODO: when we bump mathlib remove the explicit universe level
+      have me := MeasurableEquiv.measurableEmbedding e_meas
+      have hh (ξ : (i : Option ι) → Measure (β i)) [∀ (i : Option ι), IsProbabilityMeasure (ξ i)] :
+      Measure.pi ξ = Measure.map (⇑e_meas) (Measure.prod (Measure.pi fun i ↦ ξ (some i)) (ξ none))
+          := by
```

Comment:
3 remarks:
1 - indent the conclusion of the have
2 - having `:= by` alone on one line is not proper style for mathlib
3 - You can use dot notation a bit more, for example for Measure.prod
```suggestion
          Measure.pi ξ = ((Measure.pi fun i ↦ ξ (some i)).prod (ξ none)).map (⇑e_meas) := by
```
Also: do you really need the coercion arrow for `e_meas`?